### PR TITLE
Don't hold walproposer WAL in memory

### PR DESCRIPTION
**Pull request by petuhovskiy**

### CODE REVIEW

**Comment by petuhovskiy** [comment]
This PR is generally ready for review. CI tests are passing, all except `test_pageserver_catchup_while_compute_down` timeouted in debug, investigating that.

**Comment by petuhovskiy** [comment]
Looked at the test and the logs, failure is probably not related to this PR. I see that test_pageserver_catchup_while_compute_down [test is running queries on compute when pageserver is down](https://github.com/zenithdb/zenith/blob/d56a0ee19aeec715f9c839a9bcdc91c650000f1e/test_runner/batch_others/test_pageserver_catchup.py#L29-L40), seems that sometimes it's not possible and test hangs.

File: src/backend/replication/walproposer_utils.c [hunk 1]
```diff
@@ -8,6 +8,15 @@
 #include <netinet/tcp.h>
 #include <unistd.h>
 
+/*
+ * These variables are used similarly to openLogFile/SegNo,
+ * but for walproposer to write the XLOG during recovery. recvFileTLI is the TimeLineID
+ * corresponding the filename of recvFile.
+ */
+static int	recvFile = -1;
```
Comment (arssher) [inline src/backend/replication/walproposer_utils.c:16]: perhaps change prefix to 'walprop' and mention that this code is from walreceiver.c

**Review by arssher — APPROVED** [review]
Nice. So this commit removes in memory queue, puts during recovery WAL on disk in its usual place and runs separate `XLogReader` for each safekeeper. This prevents OOM when one safekeeper deeply lags behind. Let's put that into commit message. 

To preserve sending of empty `AppendRequest`s to deliver commit_lsn and other metadata to safekeepers now SendAppendRequests always sends one message even if there is no WAL data.

I added minor comments to the branch, consider merging them.


BTW I did a quick performance comparison during normal work and it is strange. With patch
test_walproposer_pgbench.tps_pgbench: 12,308.0612
without
test_walproposer_pgbench.tps_pgbench: 5,390.6744
not sure why yet

**Comment by petuhovskiy** [comment]
Performance change for a local single safekeeper setup is quite strange, but I expect some difference in a cloud setup. I'll probably try to run my cloud tests after everything is merged to main.

**Comment by petuhovskiy** [comment]
[Here are results](https://gist.github.com/petuhovskiy/ce48303382a3fbec514b8494d9b3710a) for running cloud benchmark after this patch. It's ~20k TPS, seems similar to the previous results from [here](https://github.com/neondatabase/neon/pull/1318) (~20.5k TPS).

**Pull request merged**